modules/core/04-channel/keeper/upgrade.go

```diff
@@ -68,3 +72,157 @@ func (k Keeper) constructProposedUpgrade(ctx sdk.Context, portID, channelID stri
+	proposedUpgradeFields types.UpgradeFields,
+	proposedUpgradeTimeout types.UpgradeTimeout,
+	counterpartyProposedUpgrade types.Upgrade,
+	counterpartyUpgradeSequence uint64,
+	proofCounterpartyChannel,
+	proofUpgrade []byte,
+	proofHeight clienttypes.Height,
+) (upgradeSequence uint64, err error) {
+	channel, found := k.GetChannel(ctx, portID, channelID)
+	if !found {
+		return 0, errorsmod.Wrapf(types.ErrChannelNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	// the channel state could be in INITUPGRADE if we are in a crossing hellos situation
+	if !collections.Contains(channel.State, []types.State{types.OPEN, types.INITUPGRADE}) {
+		return 0, errorsmod.Wrapf(types.ErrInvalidChannelState, "expected one of [%s, %s], got %s", types.OPEN, types.INITUPGRADE, channel.State)
+	}
+
+	// validate the proposed upgrade fields against the existing channel
+	if err = k.ValidateUpgradeFields(ctx, proposedUpgradeFields, channel); err != nil {
+		return 0, errorsmod.Wrapf(types.ErrInvalidUpgrade, "proposed upgrade fields are invalid: %s", err.Error())
+	}
+
+	connectionEnd, err := k.GetConnection(ctx, channel.ConnectionHops[0])
+	if err != nil {
+		return 0, err
+	}
+
+	counterpartyConnectionHops := []string{connectionEnd.GetCounterparty().GetConnectionID()}
+
+	expectedCounterpartyChannel := types.Channel{
+		State:           types.INITUPGRADE,
+		Counterparty:    types.NewCounterparty(channel.Counterparty.PortId, channel.Counterparty.ChannelId),
+		Ordering:        channel.Ordering,
+		ConnectionHops:  counterpartyConnectionHops,
+		Version:         channel.Version,
+		UpgradeSequence: counterpartyUpgradeSequence,
+	}
+
+	// verify that the counterparty channel has correctly entered into the upgrade process
+	if err := k.connectionKeeper.VerifyChannelState(
+		ctx,
+		connectionEnd,
+		proofHeight,
+		proofCounterpartyChannel,
+		channel.Counterparty.PortId,
+		channel.Counterparty.ChannelId,
+		expectedCounterpartyChannel,
+	); err != nil {
+		return 0, err
+	}
+
+	// verifies the proof that a particular proposed upgrade has been stored in the upgrade path.
+	if err := k.connectionKeeper.VerifyChannelUpgrade(ctx, connectionEnd, proofHeight, proofUpgrade, channel.Counterparty.PortId,
+		channel.Counterparty.ChannelId, counterpartyProposedUpgrade); err != nil {
+		return 0, err
+	}
+
+	// verify that the timeout set in UpgradeInit has not passed on this chain
+	if hasPassed, err := counterpartyProposedUpgrade.Timeout.HasPassed(ctx); hasPassed {
+		errorReceipt := types.NewErrorReceipt(channel.UpgradeSequence, err)
+		// TODO: emit error receipt events
+		k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+		return 0, errorsmod.Wrapf(types.ErrInvalidUpgrade, "upgrade timeout has passed, error receipt written for upgrade sequence: %d", channel.UpgradeSequence)
+	}
+
+	// happy path case
+	// increment upgrade sequence appropriately
+	if channel.State == types.OPEN {
+		if counterpartyUpgradeSequence > channel.UpgradeSequence {
```

Comment:
I think this comment is a little misleading, a valid crossing hello is still a happy path

```suggestion
	if channel.State == types.OPEN {
	        // increment upgrade sequence appropriately
		if counterpartyUpgradeSequence > channel.UpgradeSequence {
```